lib/vkloader.c

```diff
@@ -1044,19 +1062,36 @@ ktxTexture_VkUploadEx(ktxTexture* This, ktxVulkanDeviceInfo* vdi,
               | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT
         );
 
-        vResult = vdi->vkFuncs.vkAllocateMemory(vdi->device, &memAllocInfo,
-                                  vdi->pAllocator, &stagingMemory);
-        if (vResult != VK_SUCCESS) {
-            return KTX_OUT_OF_MEMORY;
+        uint64_t stagingAllocId = 0ull;
+        if (!useSuballocator) {
+            vResult = vdi->vkFuncs.vkAllocateMemory(vdi->device, &memAllocInfo,
+                vdi->pAllocator, &stagingMemory);
+            if (vResult != VK_SUCCESS) {
+                return KTX_OUT_OF_MEMORY;
+            }
+            VK_CHECK_RESULT(
+                    vdi->vkFuncs.vkBindBufferMemory(vdi->device, stagingBuffer,
+                        stagingMemory, 0));
+
+            VK_CHECK_RESULT(
+                    vdi->vkFuncs.vkMapMemory(vdi->device, stagingMemory, 0,
+                        memReqs.size, 0,
+                        (void**)&pMappedStagingBuffer));
+        }
+        else {
+            uint64_t numPages = 0ull;
+            stagingAllocId = subAllocatorCallbacks->allocMemFuncPtr(&memAllocInfo, &memReqs, &numPages);
+            if (stagingAllocId == 0ull) {
+                return KTX_OUT_OF_MEMORY;
+            }
+            assert(numPages == 1ull); // We do NOT support sparse bindings yet.
```

Comment:
Is numPages reliant on the provided allocmem function? If so, could it sometimes return 1 and other times > 1? If so then I think this should be an error not an assert.

Reply:
The issue here is that if the allocator has sparse binding support numPages can be > 1. Currently the way the code is structured, there needs to be additional work before the code can handle numPages > 1 (i.e. it would entail memcpy calls more than once and some heavy testing to ensure we're not messing things up). Do we have an error code like "feature not supported"?

Reply:
Just used `KTX_UNSUPPORTED_FEATURE`. Re-open if not satisfied.